File: spec/src/SixFour/Spec/MaskedBandTrainer.hs
{- |
Module      : SixFour.Spec.MaskedBandTrainer
Description : The TRAINING contract for the only learned object (θ_B) as a byte-checkable twin — pins @zero-genome == floor@, that "SixFour.Spec.MaskedBandPrediction" @trainBandJoint@ drives the masked-band loss down and RECOVERS a golden committed band, and that the descent is monotone (so the ill-conditioned trainer-divergence regime cannot pass). The golden constants ('goldenFloorBand', 'goldenTrainedBand') are what the Mac/MLX θ_B descent and the shipped 63-param forward pass must both reproduce.

The encoder is frozen ("SixFour.Spec.EncoderFrozen"); the ONLY learned object is the
63-param masked-band predictor @θ_B@. This module is its training GATE: a single fixed
training example, trained for a fixed number of steps, must take a SPECIFIC trajectory —
from the floor band to a golden recovered band — and the descent must be monotone. That
makes the MLX trainer (the @trainBandJoint@ twin) and the on-device hand-written forward
pass byte-checkable against these laws.

  * 'lawZeroGenomeIsFloor' — at the floor params the prediction is 'goldenFloorBand' (@0@)
    and the off-floor target genuinely incurs loss (the @zero-genome == floor@ start point).
  * 'lawTrainingDrivesLossDown' — 'trainBandJoint' drives the loss to a tiny fraction of the
    floor loss (the masked target IS learnable from the fixed embedding).
  * 'lawTrainedForwardIsGolden' (THE TWIN) — after training, the committed band is exactly
    'goldenTrainedBand' (@3000@). The MLX-trained θ_B and the device forward pass must both
    produce this byte; a drifted trainer or a wrong forward pass fails here.
  * 'lawTrainingDescendsMonotonically' — more steps never increase the loss. Teeth against a
    diverging / oscillating trainer (the fixed-η ill-conditioning at @ṽ → 0.9@ — kept out of
    range here by a moderate-@ṽ@ witness, and this law is the guard that would catch it).

Additive: a pure law module over "SixFour.Spec.MaskedBandPrediction"
(@trainBandJoint@/@maskedBandLoss@/@predictMaskedBand@/@zeroParamsB@). Re-pins NOTHING; no
golden vector file — the goldens are exported constants. GHC-boot-only; laws @once@-tested
in "Properties.MaskedBandTrainer".
-}
module SixFour.Spec.MaskedBandTrainer
  ( -- * The golden training fixture + its pinned trajectory endpoints
    trainerExample
  , trainerSteps
  , goldenFloorBand
  , goldenTrainedBand
    -- * The batch-STABLE trainer (mean gradient) — fixes the summed-gradient divergence
  , trainBandJointStable
    -- * Laws (closed :: Bool; @once@-tested in @Properties.MaskedBandTrainer@)
  , lawZeroGenomeIsFloor
  , lawTrainingDrivesLossDown
  , lawTrainedForwardIsGolden
  , lawTrainingDescendsMonotonically
  , lawStableTrainerSurvivesBatchDivergence
  ) where

import SixFour.Spec.MaskedBandPrediction
  ( MaskedBandExample, trainBandJoint, maskedBandLoss, maskedBandLossSum
  , maskedBandGradient, paramCountB, predictMaskedBand, zeroParamsB )

-- | The golden training fixture: coarse value @20000@ (ṽ ≈ 0.305 — moderate, well clear of
-- the ṽ → 0.9 trainer-divergence regime), masked band @0@, off-floor target @3000@, siblings
-- zero. Deterministic; the MLX trainer trains on this same example to reproduce the goldens.
trainerExample :: MaskedBandExample
trainerExample = (20000, (3000, 0, 0, 0, 0, 0, 0), 0)

-- | The fixed step count for the golden descent.
trainerSteps :: Int
trainerSteps = 2000

-- | The committed band at the floor (@zero-genome == floor@): @0@.
goldenFloorBand :: Int
goldenFloorBand = 0

-- | The committed band after training: @3000@ — the target recovered exactly through the Q16
-- crossing (@3000\/65536 · 65536 = 3000@). The byte the MLX-trained θ_B and the on-device
-- forward pass must both produce.
goldenTrainedBand :: Int
goldenTrainedBand = 3000

-- | The θ_B produced by the golden descent (shared by the laws below).
trainerTheta :: [Double]
trainerTheta = trainBandJoint trainerSteps [trainerExample]

-- | A batch-STABLE trainer: descends the MEAN per-example gradient, NOT the SUMMED gradient
-- "SixFour.Spec.MaskedBandPrediction" 'trainBandJoint' uses. The summed gradient makes the
-- effective step scale with the example count @N@: when MANY high-ṽ examples are batched the
-- @ṽ²@ feature is large, so @η·N·λ@ exceeds the GD stability bound and the descent diverges
-- to NaN ('lawStableTrainerSurvivesBatchDivergence', GHCi-verified at 8 examples). Dividing
-- the gradient by @N@ makes the step batch-size-independent and convergent. Same η = 0.2.
-- Additive: 'trainBandJoint' is unchanged, so every golden trained against it is untouched;
-- this is the trainer to use for genuine multi-example batches.
trainBandJointStable :: Int -> [MaskedBandExample] -> [Double]
trainBandJointStable n exs =
  let m       = fromIntegral (max 1 (length exs))
      meanGrad ps = map (/ m) (foldr (zipWith (+)) (replicate paramCountB 0)
                                      (map (maskedBandGradient ps) exs))
      step ps = zipWith (\p g -> p - 0.2 * g) ps (meanGrad ps)
  in iterate step zeroParamsB !! max 0 n

-- ============================================================================
-- Laws (closed predicates; @once@-tested in Properties.MaskedBandTrainer)
-- ============================================================================

-- | @zero-genome == floor@ — the descent's start point. At 'zeroParamsB' the prediction is
-- exactly 'goldenFloorBand' (@0@), and the off-floor target genuinely incurs loss (so the
-- problem is non-vacuous). Teeth: a non-zero floor prediction, or a target the floor already
-- fits, fails.
lawZeroGenomeIsFloor :: Bool
lawZeroGenomeIsFloor =
     predictMaskedBand zeroParamsB trainerExample == goldenFloorBand
  && maskedBandLoss zeroParamsB trainerExample > 1e-6

-- | Training DRIVES the loss down: after 'trainerSteps', the masked-band loss is a tiny
-- fraction (< 1e-3) of the floor loss. Teeth: a trainer that does not learn the manufactured
-- label leaves the loss near the floor and fails.
lawTrainingDrivesLossDown :: Bool
lawTrainingDrivesLossDown =
  maskedBandLoss trainerTheta trainerExample
    < 1e-3 * maskedBandLoss zeroParamsB trainerExample

-- | THE BYTE-CHECKABLE TWIN: after the golden descent, the committed band is exactly
-- 'goldenTrainedBand' (@3000@). The MLX-trained θ_B and the device hand-written forward pass
-- must both reproduce this byte. Teeth: a drifted trainer (wrong fixpoint) or a wrong forward
-- pass produces a different band and fails.
lawTrainedForwardIsGolden :: Bool
lawTrainedForwardIsGolden =
  predictMaskedBand trainerTheta trainerExample == goldenTrainedBand

-- | The descent is MONOTONE: more steps never increase the loss. Teeth against a diverging /
-- oscillating trainer — exactly the failure mode of fixed-η on an ill-conditioned input (the
-- ṽ → 0.9 regime). The moderate-ṽ golden witness stays convergent, and this law is the guard
-- that a high-ṽ fixture (or a too-large η) would fail.
lawTrainingDescendsMonotonically :: Bool
lawTrainingDescendsMonotonically =
  maskedBandLoss (trainBandJoint trainerSteps [trainerExample]) trainerExample
    <= maskedBandLoss (trainBandJoint 100 [trainerExample]) trainerExample

-- | THE BATCH-DIVERGENCE DEFECT, surfaced AND fixed (GHCi-verified). On a batch of 8 high-ṽ
-- examples the summed-gradient 'trainBandJoint' DIVERGES to NaN (η·N·λ past the stability
-- bound); the mean-gradient 'trainBandJointStable' stays FINITE and converges on the SAME
-- fixture. Teeth: this is a real, reproduced defect — the @isNaN@/non-finite witness pins it,
-- and the finite, reduced 'trainBandJointStable' loss pins the fix. (A single high-ṽ example
-- converges either way; the defect is specifically batched-N high-ṽ.) If 'trainBandJoint' is
-- ever changed to a mean gradient, the summed-divergence witness here must be revisited.
lawStableTrainerSurvivesBatchDivergence :: Bool
lawStableTrainerSurvivesBatchDivergence =
  let many       = [ (v, (3000, 0, 0, 0, 0, 0, 0), 0) | v <- [50000, 52000 .. 64000] ]  -- 8 high-ṽ
      lossSummed = maskedBandLossSum (trainBandJoint 5000 many) many
      lossStable = maskedBandLossSum (trainBandJointStable 5000 many) many
  in (isNaN lossSummed || isInfinite lossSummed || lossSummed > 1.0)  -- summed diverges/fails
     && not (isNaN lossStable || isInfinite lossStable)               -- stable stays finite
     && lossStable < 1e-3                                              -- ...and converges

File: trainer/mlx/masked_band_trainer.py
"""THE FIRST GREEN TRAINING STEP. Twin of spec/SixFour/Spec/MaskedBandTrainer.hs.

This is the byte-checkable training gate for the only learned object, theta_B. A
single fixed example, trained for a fixed number of steps from the floor, must take a
SPECIFIC trajectory and recover a golden committed band:

    trainerExample   = (20000, (3000, 0, 0, 0, 0, 0, 0), 0)    coarse 20000, mask band 0
    trainerSteps     = 2000
    goldenFloorBand  = 0      (the floor prediction, zero-genome == floor)
    goldenTrainedBand = 3000  (the byte the MLX-trained theta_B AND the device forward
                               pass must both reproduce, exactly, through the Q16 crossing)

Two trainers, both eta = 0.2 from the floor:
  * train_band_joint        - SUMMED gradient (the spec's trainBandJoint). The effective
                              step scales with batch size N, so a batch of many high-v~
                              examples drives eta*N*lambda past the GD stability bound and
                              the descent diverges to NaN.
  * train_band_joint_stable - MEAN gradient. Batch-size-independent, stays convergent.
    Both reproduce the single-example golden (summed == mean on one example).

Run this module to execute the gate. Run with `--export` to also write the trained
63-float blob that SixFour/Native/MaskedBandForward.swift loads.
"""
from __future__ import annotations

import json
import os
import sys

from jepa_loss import masked_band_gradient, masked_band_loss_sum
from theta_b import PARAM_COUNT_B, zero_params_b, predict_masked_band

# --- the golden fixture + its pinned trajectory endpoints ---
TRAINER_EXAMPLE = (20000, (3000, 0, 0, 0, 0, 0, 0), 0)
TRAINER_STEPS = 2000
GOLDEN_FLOOR_BAND = 0
GOLDEN_TRAINED_BAND = 3000
ETA = 0.2


def _zeros() -> list[float]:
    return [0.0] * PARAM_COUNT_B


def train_band_joint(n: int, exs) -> list[float]:
    """Full-batch joint training, SUMMED gradient, eta = 0.2, from the floor (spec twin)."""
    ps = zero_params_b()
    for _ in range(max(0, n)):
        acc = _zeros()
        for ex in exs:
            g = masked_band_gradient(ps, ex)
            for i in range(PARAM_COUNT_B):
                acc[i] += g[i]
        ps = [p - ETA * gi for p, gi in zip(ps, acc)]
    return ps


def train_band_joint_stable(n: int, exs) -> list[float]:
    """Full-batch joint training, MEAN gradient, eta = 0.2 (the batch-stable trainer)."""
    ps = zero_params_b()
    m = max(1, len(exs))
    for _ in range(max(0, n)):
        acc = _zeros()
        for ex in exs:
            g = masked_band_gradient(ps, ex)
            for i in range(PARAM_COUNT_B):
                acc[i] += g[i]
        ps = [p - ETA * (gi / m) for p, gi in zip(ps, acc)]
    return ps


def export_blob(path: str) -> None:
    """Write the trained 63-float theta_B + golden metadata for the device forward pass."""
    theta = train_band_joint(TRAINER_STEPS, [TRAINER_EXAMPLE])
    blob = {
        "param_count": PARAM_COUNT_B,
        "eta": ETA,
        "steps": TRAINER_STEPS,
        "trainer_example": TRAINER_EXAMPLE,
        "golden_floor_band": GOLDEN_FLOOR_BAND,
        "golden_trained_band": GOLDEN_TRAINED_BAND,
        "predicted_band": predict_masked_band(theta, TRAINER_EXAMPLE),
        "theta_b": theta,
    }
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        json.dump(blob, f, indent=2)
    print(f"exported theta_B blob -> {path} (predicted band {blob['predicted_band']})")


def _gate() -> int:
    fails = 0
    floor = zero_params_b()

    # zero-genome == floor: the start point predicts the floor band and incurs real loss.
    if predict_masked_band(floor, TRAINER_EXAMPLE) != GOLDEN_FLOOR_BAND:
        print("FAIL: floor prediction != goldenFloorBand"); fails += 1
    floor_loss = masked_band_loss_sum(floor, [TRAINER_EXAMPLE])
    if not (floor_loss > 1e-6):
        print("FAIL: floor incurs no loss (vacuous fixture)"); fails += 1

    # THE BYTE-CHECKABLE TWIN (lawTrainedForwardIsGolden): after 2000 steps -> 3000 exactly.
    theta = train_band_joint(TRAINER_STEPS, [TRAINER_EXAMPLE])
    pred = predict_masked_band(theta, TRAINER_EXAMPLE)
    if pred != GOLDEN_TRAINED_BAND:
        print(f"FAIL: trained band {pred} != goldenTrainedBand {GOLDEN_TRAINED_BAND}"); fails += 1
    else:
        print(f"  byte-exact: trained theta_B recovers band {pred} == 3000")

    # lawTrainingDrivesLossDown: trained loss < 1e-3 of the floor loss.
    trained_loss = masked_band_loss_sum(theta, [TRAINER_EXAMPLE])
    if not (trained_loss < 1e-3 * floor_loss):
        print(f"FAIL: loss not driven down ({trained_loss} vs floor {floor_loss})"); fails += 1

    # lawTrainingDescendsMonotonically: more steps never increase the loss.
    l_2000 = masked_band_loss_sum(train_band_joint(2000, [TRAINER_EXAMPLE]), [TRAINER_EXAMPLE])
    l_100 = masked_band_loss_sum(train_band_joint(100, [TRAINER_EXAMPLE]), [TRAINER_EXAMPLE])
    if not (l_2000 <= l_100):
        print(f"FAIL: descent not monotone ({l_100} -> {l_2000})"); fails += 1

    # lawStableTrainerSurvivesBatchDivergence: summed diverges on 8 high-v~; mean stays finite.
    many = [(v, (3000, 0, 0, 0, 0, 0, 0), 0) for v in range(50000, 64001, 2000)]  # 8 examples
    loss_summed = masked_band_loss_sum(train_band_joint(5000, many), many)
    loss_stable = masked_band_loss_sum(train_band_joint_stable(5000, many), many)
    summed_diverged = (loss_summed != loss_summed) or (loss_summed == float("inf")) or (loss_summed > 1.0)
    stable_finite = (loss_stable == loss_stable) and (loss_stable != float("inf"))
    if not summed_diverged:
        print(f"FAIL: summed trainer did NOT diverge on the high-v~ batch ({loss_summed})"); fails += 1
    if not (stable_finite and loss_stable < 1e-3):
        print(f"FAIL: stable trainer did not converge on the batch ({loss_stable})"); fails += 1
    if fails == 0:
        print(f"  batch guard: summed diverged ({loss_summed:.3g}), mean converged ({loss_stable:.3g})")

    print("masked_band_trainer: PASS" if fails == 0 else f"masked_band_trainer: {fails} FAIL")
    return 1 if fails else 0


if __name__ == "__main__":
    rc = _gate()
    if "--export" in sys.argv and rc == 0:
        export_blob(os.path.join(os.path.dirname(__file__), "..", "out", "theta_b_golden.json"))
    raise SystemExit(rc)
